# Data Mining
## Assignment 1 - Data Cleansing

---

## Aadam - CS1945

---

## Tasks to do
1. Combine all the three datasets in a single source, you may use any DBMS. 
2. Once you have the data in the staging area inside your DBMS, perform data profiling for all the fields. By data profiling means the following statistics:
    - No. of unique values (for each column) 
    - No. of nulls (for each column) 
    - Invalid values (for each column) 
    - Total no. of courses 
    - Total no. of female students Vs male students 
    - Total no of people who has taken more than 5 courses in a semester. 
    - The relationship between total no. of unique student ID’s and total no. of students. 
    - Average no of people per semester of each campus. 
    - Average no of students in every batch of each campus.
3. After the profiling, you should have identified the anomalies and data cleansing issues. Here are some of the issues you need to address during your cleansing work.
    - Separate first and last names both for the student and father. Standardize the first and last names. (Hint: Find all the unique names in data and create a lookup table with two columns i.e. correct_name, variation etc. Use the lookup table to update the name fields with standardized names. Never hardcode names in your SQL.)
    - Introduce a new column i.e. city_name. It will involve extracting the city from the address and then the standardization of the city names. (Use the city names in the telephone directory as standard)
    - Bring the gender information into a consistent representation. Use ‘M’ and ‘F’ with data type CHAR(1).
    - Since the gender information is missing for the Peshawar campus, you can use the student names to figure out the gender. (Hint: Find all distinct male and female names and create a lookup table)
    - If you have been using VARCHAR or CHAR for date, port the date information into columns with the proper Date data type. The dates should be as per calendar dates.
    - Validate all the dates against the business rules e.g. the DOB should be smaller than the Reg. Date and Graduation Date should be greater than Reg. Date. The data might be having anomalies like exchanged DOB and Reg. Date by mistake etc. Also be careful with invalid dates i.e. 31st Feb. or 29th Feb. in a non-leap year.
    - For some campuses, the degree information is missing. Devise some technique to figure it out and update the rows with empty degree fields. Validate other businessrules for each field. One example is that marks should be in the range 0 to 100 inclusive.


In [2]:
import sqlite3

import pandas as pd

In [120]:
con = sqlite3.connect('University.db')

df_std = pd.read_sql_query("SELECT * FROM Students", con)
df_reg = pd.read_sql_query("SELECT * FROM Course_Registrations", con)

con.close()

In [116]:
df_std.head()

SID                 Name                Father  \
0     KHR_BS_0       Hussain Ansary          Ubaid Ansary   
1     KHR_BS_1  Shk. Munir Hussaini  Viqar Hamid Hussaini   
2    KHR_BS_10         Hamna Ansary         Hameed Ansary   
3   KHR_BS_100        Jabbar Haqqie     Muneer Rai Haqqie   
4  KHR_BS_1000     Rana Hayyat Baig   Ghulam Mustafa Baig   

                   DoB Gender   Reg Date Reg Status Degree Status  \
0  1974-01-05 00:00:00      M 1994-08-13          A             C   
1  1974-12-13 00:00:00      M 1994-08-13          A             C   
2  1974-04-25 00:00:00      F 1994-08-14          A             C   
3  1974-05-22 00:00:00      M 1994-08-18          A             C   
4  1976-02-25 00:00:00      M 1995-08-28          A             C   

                                             Address Qualification Degree  \
0           h# 978 Street No.72  Defense Phase 1 KHR       A-Level     BS   
1                    H#  169 s#  0 Mutian wala Thata       A-Level     BS   
2                     h#  697 St. #  94 sea site KHR       A-Level     BS   
3                       H# 504 S No.4  Kumria Quetta       A-Level     BS   
4  H No.509 St. No.38  Kishwar heights university...          HSSC     BS   

  Campus  
0    KHR  
1    KHR  
2    KHR  
3    KHR  
4    KHR

In [121]:
df_reg.head()

SID  Course  Score Semester                        Year Discipline  \
0  KHR_BS_0  CS-101     72     Fall  1994-01-01 00:00:00.000000         CS   
1  KHR_BS_0  CS-102     78     Fall  1994-01-01 00:00:00.000000         CS   
2  KHR_BS_0  CS-103     53     Fall  1994-01-01 00:00:00.000000         CS   
3  KHR_BS_0  CS-104     58     Fall  1994-01-01 00:00:00.000000         CS   
4  KHR_BS_0  CS-105     61   Spring  1995-01-01 00:00:00.000000         CS   

  Degree Campus  
0     BS    KHR  
1     BS    KHR  
2     BS    KHR  
3     BS    KHR  
4     BS    KHR

# Data Profiling

In [117]:
df_std.describe(include='all')

SID           Name            Father  \
count         17102          17102             17102   
unique        17102          14673             14270   
top     PEW_BS_1573  Kishwar Pasha  Mudassir Khattak   
freq              1              6                 5   
first           NaN            NaN               NaN   
last            NaN            NaN               NaN   

                               DoB Gender             Reg Date Reg Status  \
count                        17100  13398                17102      17102   
unique                        7380      2                  442          2   
top     1981-03-22 00:00:00.000000      M  2003-08-16 00:00:00          A   
freq                             8   8935                   76      17100   
first                          NaN    NaN  1994-01-08 00:00:00        NaN   
last                           NaN    NaN  2005-01-12 00:00:00        NaN   

       Degree Status                                 Address Qualification  \
count          17102                                   17102         17102   
unique             2                                   15360            12   
top                C  house no.141 S# 16  Haider road Multan         F.Sc.   
freq           12500                                       3          2390   
first            NaN                                     NaN           NaN   
last             NaN                                     NaN           NaN   

       Degree Campus  
count   17102  17102  
unique      2      3  
top        BS    KHR  
freq    14302   8201  
first     NaN    NaN  
last      NaN    NaN

In [122]:
df_reg.describe(include='all')

SID  Course          Score Semester  \
count        406200  406200  406200.000000   406200   
unique        11700     120            NaN        2   
top     PEW_BS_1573  CS-103            NaN   Spring   
freq             48    8900            NaN   203100   
mean            NaN     NaN      74.460268      NaN   
std             NaN     NaN      14.415844      NaN   
min             NaN     NaN      50.000000      NaN   
25%             NaN     NaN      62.000000      NaN   
50%             NaN     NaN      74.000000      NaN   
75%             NaN     NaN      87.000000      NaN   
max             NaN     NaN      99.000000      NaN   

                              Year Discipline  Degree  Campus  
count                       406200     406200  406200  406200  
unique                          16          8       2       3  
top     1970-01-01 00:00:00.000002         TC      BS     LHR  
freq                         76200      94200  376800  190800  
mean                           NaN        NaN     NaN     NaN  
std                            NaN        NaN     NaN     NaN  
min                            NaN        NaN     NaN     NaN  
25%                            NaN        NaN     NaN     NaN  
50%                            NaN        NaN     NaN     NaN  
75%                            NaN        NaN     NaN     NaN  
max                            NaN        NaN     NaN     NaN

In [124]:
df_std.nunique().to_frame()

0
SID            17102
Name           14673
Father         14270
DoB             7380
Gender             2
Reg Date         442
Reg Status         2
Degree Status      2
Address        15360
Qualification     12
Degree             2
Campus             3

In [123]:
df_reg.nunique().to_frame()

0
SID         11700
Course        120
Score          50
Semester        2
Year           16
Discipline      8
Degree          2
Campus          3

## No. of Unique Values in Students Table

| Column        | Count |
|:--------------|------:|
| SID           | 17102 |
| Name          | 14673 |
| Father        | 14270 |
| DoB           |  7380 |
| Gender        |     2 |
| Reg Date      |   442 |
| Reg Status    |     2 |
| Degree Status |     2 |
| Address       | 15360 |
| Qualification |    12 |
| Degree        |     2 |
| Campus        |     3 |

## No. of Unique Values in Course Registration Table

|            |Count |
|:-----------|-----:|
| SID        | 8000 |
| Course     |  120 |
| Score      |   50 |
| Semester   |    2 |
| Year       |   13 |
| Discipline |    8 |
| Degree     |    2 |


In [125]:
df_std.isnull().sum(axis = 0)

SID                 0
Name                0
Father              0
DoB                 2
Gender           3704
Reg Date            0
Reg Status          0
Degree Status       0
Address             0
Qualification       0
Degree              0
Campus              0
dtype: int64

In [126]:
df_reg.isnull().sum(axis = 0)

SID           0
Course        0
Score         0
Semester      0
Year          0
Discipline    0
Degree        0
Campus        0
dtype: int64

## No. of Nulls in Students Table


|               |Count |
|:--------------|-----:|
| SID           |    0 |
| Name          |    0 |
| Father        |    0 |
| DoB           |    2 |
| Gender        | 3704 |
| Reg Date      |    0 |
| Reg Status    |    0 |
| Degree Status |    0 |
| Address       |    0 |
| Qualification |    0 |
| Degree        |    0 |
| Campus        |    0 |

## No. of Nulls in Course Registration Table

|            |   0 |
|:-----------|----:|
| SID        |   0 |
| Course     |   0 |
| Score      |   0 |
| Semester   |   0 |
| Year       |   0 |
| Discipline |   0 |
| Degree     |   0 |

## Invalid Values

Let's try to see if the dates in DoB and Reg Date are valid.

In [46]:
dob = pd.to_datetime(df_std.DoB, errors="coerce")

In [127]:
df_std["Reg Date"] = pd.to_datetime(df_std["Reg Date"])

In [128]:
df_reg.Year = pd.to_datetime(df_reg.Year)

### Following are the invalid Date values in DOB

In [198]:
# df_std[dob.isnull()]

Most of the values have 29 Feb as a date in an year which wasn't a **Leap Year**. We can correct them by decrementing the day.

In [49]:
sum(dob.isnull())

16

## No. of Invalid Dates: 16

In [129]:
df_reg.Course.nunique()

120

## Total No. of Courses: 120

In [53]:
num_male = sum(df_std.Gender == 'M')
num_male

8935

In [54]:
num_female = sum(df_std.Gender == 'F')
num_female

4463

## Total no. of Male vs Female students

|    Gender     |Count |
|:--------------|-----:|
| Male          | 8935 |
| Female        | 4463 |

This doesn't contain the data from Peshawar Campus.

In [130]:
df_reg.head()

SID  Course  Score Semester       Year Discipline Degree Campus
0  KHR_BS_0  CS-101     72     Fall 1994-01-01         CS     BS    KHR
1  KHR_BS_0  CS-102     78     Fall 1994-01-01         CS     BS    KHR
2  KHR_BS_0  CS-103     53     Fall 1994-01-01         CS     BS    KHR
3  KHR_BS_0  CS-104     58     Fall 1994-01-01         CS     BS    KHR
4  KHR_BS_0  CS-105     61   Spring 1995-01-01         CS     BS    KHR

In [89]:
grouped_semester_course = df_reg.groupby(['SID', 'Semester', 'Year']).agg({'Course': ['count']})
grouped_semester_course.columns = ['count']
grouped_semester_course = grouped_semester_course.reset_index()

In [90]:
sum(grouped_semester_course['count'] > 5)

41200

## Total No. of people who has taken more than 5 courses in a semester: 41200

In [93]:
df_std.SID.count()

17102

In [101]:
df_std.SID.str.extract('(\d+)').nunique()

0    6601
dtype: int64

## The relationship between total no. of unique student ID’s and total no. of students

|         |Count |
|:--------------|-----:|
| Unique IDs          | 6601 |
| Total Students      | 17102 |



In [132]:
df_reg.head()

SID  Course  Score Semester       Year Discipline Degree Campus
0  KHR_BS_0  CS-101     72     Fall 1994-01-01         CS     BS    KHR
1  KHR_BS_0  CS-102     78     Fall 1994-01-01         CS     BS    KHR
2  KHR_BS_0  CS-103     53     Fall 1994-01-01         CS     BS    KHR
3  KHR_BS_0  CS-104     58     Fall 1994-01-01         CS     BS    KHR
4  KHR_BS_0  CS-105     61   Spring 1995-01-01         CS     BS    KHR

In [147]:
grouped_semester_students = df_reg.groupby(['Campus', 'Semester', 'Year']).agg({'SID': ['count']})
grouped_semester_students.columns = ['count']
grouped_semester_students = grouped_semester_students.reset_index()
grouped_semester_students.head()

Campus Semester       Year  count
0    KHR     Fall 1994-01-01   3600
1    KHR     Fall 1995-01-01   7200
2    KHR     Fall 1996-01-01   7200
3    KHR     Fall 1997-01-01   7200
4    KHR     Fall 1998-01-01   3600

In [146]:
avg_semester_students = grouped_semester_students.reset_index().groupby(['Campus']).agg({'count': ['mean']})
avg_semester_students.columns = ['avg']
avg_semester_students = avg_semester_students.reset_index()
avg_semester_students

Campus           avg
0    KHR   4133.333333
1    LHR   7950.000000
2    PEW  28200.000000

## Average no of people per semester of each campus

| Campus   |      Avg |
|:---------|---------:|
| KHR      |  4133.33 |
| LHR      |  7950    |
| PEW      | 28200    |

# Data Cleansing

---

- Introduce a new column i.e. city_name. It will involve extracting the city from the address and then the standardization of the city names. (Use the city names in the telephone directory as standard)

In [167]:
city_names = df_std.Address.str.split().str[-1]

In [219]:
last_two_city_names = df_std.Address.str.split().str[-2:].str.join(' ')
# sorted(last_two_city_names.unique())

In [220]:
sorted(city_names.unique())

['Abbotabad',
 'Attock',
 'Bahawalpur',
 'Bhatian',
 'Bhawalpur',
 'Chishtian',
 'D.G. Khan',
 'D.I. Khan',
 'Faisalabad',
 'Gilgit',
 'Gujrat',
 'Gujrawala',
 'Hakim',
 'Haripur',
 'Hasanabdal',
 'Hyderabad',
 'Ismail Khan',
 'Jehlum',
 'KARACHI',
 'KHR',
 'Kabirwala',
 'Karachi',
 'Kharian',
 'Khas',
 'Khr',
 'LalaMusa',
 'Lyyah',
 'Mansehra',
 'Mardan',
 'Melsi',
 'Multan',
 'Noshera',
 'PESHAWAR',
 'PSH',
 'Peshawar',
 'Psh',
 'Quetta',
 'Sadiqabad',
 'Sahiwal',
 'Sargodha',
 'Sawat',
 'Shekhopura',
 'Sialkot',
 'Sibbi',
 'Thata',
 'Umer Kot',
 'Usman Kot',
 'Wahari',
 'Wazirabad',
 'ghazi khan',
 'karachi',
 'khr',
 'peshawar',
 'psh']

In [239]:
std_city_map = {}
std_city_map.update(
    dict.fromkeys(
        [
            'KHR', 
            'KARACHI', 
            'Khr', 
            'Karachi', 
            'khr', 
            'karachi'
            ], 'Karachi'))

std_city_map.update(
    dict.fromkeys(
        [
            'PSH',
            'Psh',
            'Peshawar',
            'PESHAWAR',
            'peshawar',
            'psh'
        ], 'Peshawar'
    )
)

std_city_map.update({'Abbotabad': 'Abottabad'})
std_city_map.update({'Attock': 'Attock'})
std_city_map.update({'Bahawalpur': 'Bahawalpur', 'Bhawalpur': 'Bahawalpur'})
std_city_map.update({'Bhatian': 'Pindi Bhattian'})
std_city_map.update({'Chishtian': 'Chishtian'})
std_city_map.update({'D.G. Khan': 'Dera Ghazi Khan', 'ghazi khan': 'Dera Ghazi Khan'})
std_city_map.update({'D.I. Khan': 'D.I.Khan', 'Ismail Khan': 'D.I.Khan'})
std_city_map.update({'Faisalabad': 'Faisalabad'})
std_city_map.update({'Gilgit': 'Gilgit'})
std_city_map.update({'Gujrat': 'Gujrat'})
std_city_map.update({'Gujrawala': 'Gujranwala'})
std_city_map.update({'Hakim': 'Abdul Hakim'})
std_city_map.update({'Haripur': 'Haripur'})
std_city_map.update({'Hasanabdal': 'Hasan Abdal'})
std_city_map.update({'Hyderabad': 'Hyderabad'})
std_city_map.update({'Jehlum': 'Jhelum'})
std_city_map.update({'Kabirwala': 'Kabirwala'})
std_city_map.update({'Kharian': 'Kharian'})
std_city_map.update({'Khas': 'Mirpur Khas'})
std_city_map.update({'LalaMusa': 'LalaMusa'})
std_city_map.update({'Lyyah': 'Layyah'})
std_city_map.update({'Mansehra': 'Mansehra'})
std_city_map.update({'Mardan': 'Mardan'})
std_city_map.update({'Melsi': 'Mailsi'})
std_city_map.update({'Noshera': 'Nowshera'})
std_city_map.update({'Sawat': 'Swat'})
std_city_map.update({'Shekhopura': 'Sheikhupura'})
std_city_map.update({'Sibbi': 'Sibi'})
std_city_map.update({'Thata': 'Thatta'})
std_city_map.update({'Wahari': 'Vehari'})

In [253]:
# std_city_map

In [241]:
two_words_city_name = ['Khan', 'Kot']

In [254]:
# Get the previous word for these, in order to figure out the city
city_names[city_names == 'Khan'] = last_two_city_names[city_names == 'Khan']
city_names[city_names == 'khan'] = last_two_city_names[city_names == 'khan']
city_names[city_names == 'Kot'] = last_two_city_names[city_names == 'Kot']

In [256]:
# sorted(city_names.unique())

In [257]:
std_city_names = city_names.replace(std_city_map)

In [258]:
sorted(std_city_names.unique())

['Abdul Hakim',
 'Abottabad',
 'Attock',
 'Bahawalpur',
 'Chishtian',
 'D.I.Khan',
 'Dera Ghazi Khan',
 'Faisalabad',
 'Gilgit',
 'Gujranwala',
 'Gujrat',
 'Haripur',
 'Hasan Abdal',
 'Hyderabad',
 'Jhelum',
 'Kabirwala',
 'Karachi',
 'Kharian',
 'LalaMusa',
 'Layyah',
 'Mailsi',
 'Mansehra',
 'Mardan',
 'Mirpur Khas',
 'Multan',
 'Nowshera',
 'Peshawar',
 'Pindi Bhattian',
 'Quetta',
 'Sadiqabad',
 'Sahiwal',
 'Sargodha',
 'Sheikhupura',
 'Sialkot',
 'Sibi',
 'Swat',
 'Thatta',
 'Umer Kot',
 'Usman Kot',
 'Vehari',
 'Wazirabad']

In [259]:
df_std['City'] = std_city_names

In [260]:
df_std.head()

SID                 Name                Father  \
0     KHR_BS_0       Hussain Ansary          Ubaid Ansary   
1     KHR_BS_1  Shk. Munir Hussaini  Viqar Hamid Hussaini   
2    KHR_BS_10         Hamna Ansary         Hameed Ansary   
3   KHR_BS_100        Jabbar Haqqie     Muneer Rai Haqqie   
4  KHR_BS_1000     Rana Hayyat Baig   Ghulam Mustafa Baig   

                   DoB Gender   Reg Date Reg Status Degree Status  \
0  1974-01-05 00:00:00      M 1994-08-13          A             C   
1  1974-12-13 00:00:00      M 1994-08-13          A             C   
2  1974-04-25 00:00:00      F 1994-08-14          A             C   
3  1974-05-22 00:00:00      M 1994-08-18          A             C   
4  1976-02-25 00:00:00      M 1995-08-28          A             C   

                                             Address Qualification Degree  \
0           h# 978 Street No.72  Defense Phase 1 KHR       A-Level     BS   
1                    H#  169 s#  0 Mutian wala Thata       A-Level     BS   
2                     h#  697 St. #  94 sea site KHR       A-Level     BS   
3                       H# 504 S No.4  Kumria Quetta       A-Level     BS   
4  H No.509 St. No.38  Kishwar heights university...          HSSC     BS   

  Campus     City  
0    KHR  Karachi  
1    KHR   Thatta  
2    KHR  Karachi  
3    KHR   Quetta  
4    KHR  Karachi

In [261]:
df_std.describe()

SID           Name            Father  \
count         17102          17102             17102   
unique        17102          14673             14270   
top     PEW_BS_1573  Kishwar Pasha  Mudassir Khattak   
freq              1              6                 5   
first           NaN            NaN               NaN   
last            NaN            NaN               NaN   

                               DoB Gender             Reg Date Reg Status  \
count                        17100  13398                17102      17102   
unique                        7380      2                  442          2   
top     1981-03-22 00:00:00.000000      M  2003-08-16 00:00:00          A   
freq                             8   8935                   76      17100   
first                          NaN    NaN  1994-01-08 00:00:00        NaN   
last                           NaN    NaN  2005-01-12 00:00:00        NaN   

       Degree Status                                 Address Qualification  \
count          17102                                   17102         17102   
unique             2                                   15360            12   
top                C  house no.141 S# 16  Haider road Multan         F.Sc.   
freq           12500                                       3          2390   
first            NaN                                     NaN           NaN   
last             NaN                                     NaN           NaN   

       Degree Campus     City  
count   17102  17102    17102  
unique      2      3       41  
top        BS    KHR  Karachi  
freq    14302   8201     4438  
first     NaN    NaN      NaN  
last      NaN    NaN      NaN

### We have 41 Unique cities

---

- Bring the gender information into a consistent representation. Use ‘M’ and ‘F’ with data type CHAR(1).


We've already achieved this during the data collection phase.